spanner_dbapi/connection.py

```diff
@@ -5,87 +5,168 @@
+    @wraps(func)
+    def wrapped(self, *args, **kwargs):
+        if self._is_closed:
+            raise InterfaceError("connection is already closed")
+        return func(self, *args, **kwargs)
+
+    return wrapped
+
+
+class Connection(object):
+    """This is a wrap-around object for the existing `Database` and the
+    corresponding `Instance` objects.
+
+    :type database: :class:`~google.cloud.spanner_v1.database.Database`
+    :param database: Corresponding Database
+
+    :type instance: :class:`~google.cloud.spanner_v1.instance.Instance`
+    :param instance: The instance that owns the database.
+
+    :type autocommit: bool
+    :param autocommit: (Optional) When changed to True, all the pending
+                       transactions must be committed.
+
+    :type read_only: bool
+    :param read_only: (Optional) Indicates if the Connection is intended to be
+                      used for read-only transactions.
+    """
+    def __init__(self, database, instance, autocommit=True, read_only=False):
+        self.database = database
+        self.instance = instance
+        self.autocommit = autocommit
+        self.read_only = read_only
+        self.transaction_mode = (
+            TransactionModes.read_only
+            if self.read_only
+            else TransactionModes.read_write
+        )
+        self.autocommit_dml_mode = AutocommitDMLModes.transactional
+        self.timeout_secs = 0
+        self.read_timestamp = None
+        self.commit_timestamp = None
+        self._is_closed = False
+        self._inside_transaction = not autocommit
+        self._transaction_started = False
         self._ddl_statements = []
+        self.read_only_staleness = {}
 
-    def cursor(self):
-        self.__raise_if_already_closed()
+    @property
+    def is_closed(self):
+        return self._is_closed
 
-        return Cursor(self)
+    @property
+    def inside_transaction(self):
+        return self._inside_transaction
 
-    def __raise_if_already_closed(self):
-        """
-        Raise an exception if attempting to use an already closed connection.
-        """
-        if self._closed:
-            raise InterfaceError("connection already closed")
+    @property
+    def transaction_started(self):
+        return self._transaction_started
 
-    def __handle_update_ddl(self, ddl_statements):
-        """
-        Run the list of Data Definition Language (DDL) statements on the underlying
-        database. Each DDL statement MUST NOT contain a semicolon.
-        Args:
-            ddl_statements: a list of DDL statements, each without a semicolon.
-        Returns:
-            google.api_core.operation.Operation.result()
+    @property
+    def ddl_statements(self):
+        return self._ddl_statements
+
+    @ddl_statements.setter
+    def ddl_statements(self, dll):
+        self._change_transaction_started(True)
+
+        self._ddl_statements = dll
+        if self.autocommit:
+            self.commit()
```

Comment:
Not sure if this is a good way of use: user changes the `ddl_statements` value, and it becomes commited in the same moment. It probably should be moved to `append_ddl_statements()`, while setting `ddl_statements` like "= value" should be disabled.